peps/pep-0740.rst

```diff
@@ -205,110 +181,346 @@ changes to it:
 
 .. _provenance-object:
 
 Provenance objects
-^^^^^^^^^^^^^^^^^^
+------------------
 
 The index will serve uploaded attestations along with metadata that can assist
 in verifying them in the form of JSON serialized objects.
 
-These "provenance objects" will be available via both the :pep:`503` Simple Index
-and :pep:`691` JSON-based Simple API as described below, and will have the
-following structure:
+These *provenance objects* will be available via both the :pep:`503` Simple Index
+and :pep:`691` JSON-based Simple API as described above, and will have the
+following layout:
 
 .. code-block:: json
 
     {
-        "publisher": {
-          "type": "important-ci-service",
-          "claims": {},
-          "vendor-property": "foo",
-          "another-property": 123
-        },
-        "attestations": {
-          "some-attestation": {/* ... */},
-          "another-attestation": {/* ... */}
-        }
+        "version": 1,
+        "attestation_bundles": [
+          {
+            "publisher": {
+              "kind": "important-ci-service",
+              "claims": {},
+              "vendor-property": "foo",
+              "another-property": 123
+            },
+            "attestations": [
+              { /* attestation 1 ... */ },
+              { /* attestation 2 ... */ }
+            ]
+          }
+        ]
     }
 
-* ``publisher`` is an **optional** JSON object, containing a
-  representation of the file's Trusted Publisher configuration at the time
-  the file was uploaded to the package index. The keys within the ``publisher``
-  object are specific to each Trusted Publisher but include, at minimum:
-
-  * A ``type`` key, which **MUST** be a JSON string that uniquely identifies the
+or, as pseudocode:
+
+.. code-block:: python
+
+  @dataclass
+  class Publisher:
+      kind: string
+      """
+      The kind of Trusted Publisher.
+      """
+
+      claims: object | None
+      """
+      Any context-specific claims retained by the index during Trusted Publisher
+      authentication.
+      """
+
+      _rest: object
+      """
+      Each publisher object is open-ended, meaning that it MAY contain additional
+      fields beyond the ones specified explicitly above. This field signals that,
+      but is not itself present.
+      """
+
+  @dataclass
+  class AttestationBundle:
+      publisher: Publisher
+
+      attestations: list[Attestation]
+
+  @dataclass
+  class Provenance:
+      version: Literal[1]
+      """
+      The provenance object's version, which is always 1.
+      """
+
+      attestation_bundles: list[AttestationBundle]
+      """
+      One or more attestation "bundles".
+      """
+
+* ``version`` is ``1``. Like attestation objects, provenance objects are
+  versioned, and this PEP only defines version ``1``.
+* ``attestation_bundles`` is a **required** JSON array, containing one
+  or more "bundles" of attestations, paired with their signing identity
+  (such as a Trusted Publishing identity).
```

Comment:
Another possible wording:

```suggestion
* ``attestation_bundles`` is a **required** JSON array, containing one
  or more "bundles" of attestations. Each bundle corresponds to a
  signing identity (such as a Trusted Publishing identity), and contains
  one or more attestation objects.
```

To try to make clear that besides having multiple bundles, each bundle can have multiple attestations and the "bundle" is a way of grouping them by signing identity